### DISCUSSION TITLE
How to change environment variables ;)

### DISCUSSION THREAD

**Discussion by JC5** [firefly-iii/firefly-iii#8332] [General]
So we don't pollute [#6312](https://github.com/firefly-iii/firefly-iii/issues/6312).

@maureenferreira, it's a complete overhaul. You can try it already by setting `FIREFLY_III_LAYOUT=v2`.

**Comment by terrencetingleberry** [comment]
Alright, thanks! I'm very excited to try it out... how far along would you say it is (if you have a rough estimate)? Also, perhaps you could add a link to this discussion in the description of #6312, it might be helpful.

**Comment by terrencetingleberry** [comment]
Also, where exactly do I go to set the `FIREFLY_III_LAYOUT` tag to `v2`?

**Reply by JC5** [reply]
Wherever you set your environment variables like `APP_KEY` and the rest.

**Reply by terrencetingleberry** [reply]
Alright, I'll try finding it.

**Reply by terrencetingleberry** [reply]
My apologies if this should be obvious, but I've tried to find where to edit the environment variables for my Firefly Docker container, but I can't seem to find it anywhere. I also tried searching for the answer online to no avail. Where exactly do I go in order to set environment variables for containers?

**Reply by JC5** [reply]
Eh, where you set them last time?

**Reply by terrencetingleberry** [reply]
Oh, okay! I thought those were only for pre-build. Thanks!

**Reply by terrencetingleberry** [reply]
@JC5 So, after much research, I've learned that, as far as I could find, you can't change environment variables in an already existing Docker container. Once you set the environment variables, build an image, and deploy it, the environment variables are set in stone. Correct me if I'm wrong.

**Reply by JC5** [reply]
Alright let's take a step back and rename this thread lol, we're getting a little side-tracked here.

Docker compose: if you run Firefly III using docker compose, you can edit environment variables inside the docker-compose.yml file or edit the file referred to by `env_file`. This is usually `.env`. The key pairs you see can be edited as you wish.

Docker: if you run Firefly III on the command line you can edit the command you gave to start the container and add or edit environment variables by using `-e FOO=bar`.

Portainer: either edit the variables in the stack or edit stack.env 

https://docs.firefly-iii.org/how-to/firefly-iii/installation/docker/

**Reply by terrencetingleberry** [reply]
Thanks for the response! I didn't think this was the right method, as I already tried that previously. So, I tried editing `FIREFLY_III_LAYOUT` in the `.env` file once more and this time restarting my Docker container after doing so. However, the UI is still the same.

**Reply by JC5** [reply]
That's weird. If you set it to `FIREFLY_III_LAYOUT=v2` and you're running the latest version, it should look something like this:

<img width="2093" alt="Screenshot 2024-01-06 at 14 49 08" src="https://github.com/firefly-iii/firefly-iii/assets/5889984/e5411a99-5291-4784-8ea0-bca41be95ea3">

**Reply by terrencetingleberry** [reply]
Here's what my dashboard looks like:
![image](https://github.com/firefly-iii/firefly-iii/assets/142938968/ef8894f1-e8f9-40c7-aab1-0bb3f9449ec1)

**Reply by JC5** [reply]
Mmm, that's surprising, I see nothing. You're using Docker, right? Docker compose?

**Reply by terrencetingleberry** [reply]
> You're using Docker, right? Docker compose?

Yep. I followed the standard tutorial for Firefly with the Data Importer. I've actually done it twice now just to make absolutely sure I didn't do anything wrong, and I'm sure I didn't. I followed the tutorial to a T, and everything appears normally when I check in the Docker Hub app, run `docker ps`, and so on. If I were to rebuild the images with the `FIREFLY_III_LAYOUT` var set to `v2`, I can almost surely promise you it would have the new UI enabled, the issue is just that none of the variables I edit after the images are built seem to take effect.

Perhaps I should move this discussion to an issue?

**Reply by marcelbrueckner** [reply]
Assuming you're using a compose file based on https://raw.githubusercontent.com/firefly-iii/docker/main/docker-compose.yml, you're storing your environment variables in an `.env` file (opposed to storing them directly in the compose file under `env:` key).

> So, I tried editing FIREFLY_III_LAYOUT in the .env file once more and this time restarting my Docker container after doing so.

When changing an environment variable, the **container needs to be re-created**. Just restarting the container isn't sufficient. After you've made the changes in your `.env` file, you need to run `docker compose up -d` (or `docker-compose` with dash, depending on your Docker version). In the end, it should look like this.

```bash
$ sudo docker-compose up -d
firefly-iii-db is up-to-date
firefly-iii-data-importer is up-to-date
Recreating firefly-iii ... done
```

**Reply by terrencetingleberry** [reply]
Ah, alright! Thanks, I'll try that.
